### PR TITLE
Refactored the PipelineParser/PipelineConfiguration to use the PipelineDataFlowModel

### PR DESCRIPTION
**Pull request by dlvenable**
### Description

Prior to this PR, Data Prepper had split approaches for serializing configuration YAML from deserializing YAML. The `PipelineDataFlowModel` is currently being used for serialization. And deserialization is handled by the `PipelineParser` and `PipelineConfiguration` classes. This PR fixes this split-brain issue. Deserialization now uses `PipelineDataFlowModel` as well. This helps ensure that deserialization is consistent with serialization.

This change also starts to remove the legacy `prepper:` configuration in YAML. I wasn't intending to do this, but it would be harder to keep that behavior and we plan to remove it in #619.

This change is important for supporting conditional routing of sinks.
 
### Issues Resolved

None, but contributes toward #619 and #1337 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java [hunk 1]
```diff
@@ -5,36 +5,53 @@
...
-        final PipelineConfiguration pipelineConfiguration = new PipelineConfiguration(TestDataProvider.validSingleConfiguration(),
-                null,
-                TestDataProvider.validMultipleConfigurationOfSizeOne(),
-                TestDataProvider.validMultipleConfiguration(),
-                TestDataProvider.TEST_WORKERS, TestDataProvider.TEST_DELAY);
+    void testPipelineConfigurationCreation() {
+        final PipelineModel pipelineModel = mock(PipelineModel.class);
+        when(pipelineModel.getSource()).thenReturn(source);
+        when(pipelineModel.getProcessors()).thenReturn(processors);
+        when(pipelineModel.getPreppers()).thenReturn(null);
```
Comment (sshivanii) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java:49]: Do we still `preppers` for this when statement?

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java [hunk 2]
```diff
@@ -67,78 +84,16 @@ public void testPipelineConfigurationCreation() {
...
-
-    @Test
-    public void testOnlySourceAndSink() {
-        final PipelineConfiguration pipelineConfiguration = new PipelineConfiguration(TestDataProvider.validSingleConfiguration(),
-                null,
-                null,
-                TestDataProvider.validMultipleConfigurationOfSizeOne(),
-                null, null);
+    void testOnlySourceAndSink() {
+        sinks = TestDataProvider.validMultipleConfigurationOfSizeOne();
```
Comment (sshivanii) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java:88]: Why are we not continuing to use `validMultipleConfiguration` for `sinks`?

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java [hunk 3]
```diff
@@ -155,108 +110,102 @@ public void testOnlySourceAndSink() {
...
-                    TestDataProvider.validSingleConfiguration(),
-                    TestDataProvider.validMultipleConfiguration(),
-                    TestDataProvider.validMultipleConfiguration(),
-                    TestDataProvider.TEST_WORKERS, TestDataProvider.TEST_DELAY);
-        } catch (IllegalArgumentException ex) {
-            assertThat(ex.getMessage(), is("Invalid configuration, source is a required component"));
-        }
+    @Test
+    void testNoSourceConfiguration() {
+        final PipelineModel pipelineModel = mock(PipelineModel.class);
```
Comment (sshivanii) [inline data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java:115]: Don't we want to pass `null` for source here?

**Review by sshivanii — COMMENTED** [review]
Overall, the serialization/deserialization consistency by using `PipelineDataFlowModel` is a good change. I just have few minor clarifying comments.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java:49]: I might be able to remove this line. I originally had to set this to `null` for preppers to work. But, then some other issues came up, so I removed any support for prepper: in the PipelineConfiguration class.

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java:115]: Mockito's default behavior is to provide `null` when not mocked (it is different for collections and primitives).

File: data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java (hunk 2 above)
Comment (dlvenable) [reply data-prepper-core/src/test/java/org/opensearch/dataprepper/parser/model/PipelineConfigurationTests.java:88]: I don't think it matters. But, the original test was using size of one. So, I kept it that way. I wanted to keep the tests as similar as possible to their original intent.

**Review by sshivanii — APPROVED** [review]

**Review by asifsmohammed — APPROVED** [review]

**Pull request merged**